### ISSUE TITLE
pvget crash w/ "Timer::cancel() onList==true, but not found"

### ISSUE THREAD

**Issue by mdavidsaver** [epics-base/pvAccessCPP#152]
I had an instance of pvget crashing with an unhandled exception:

```
terminate called after throwing an instance of 'std::logic_error'
  what():  Timer::cancel() onList==true, but not found
```

This indicates corruption on a timer queue.  From the core dump, this appears to happen during TCP connection shutdown.  The stack trace doesn't match up with the exception though, so this may be a side effect of previous corruption.

```c++
Thread 1 (Thread 0x7fe99cb55700 (LWP 27898)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fe99e6b4535 in __GI_abort () at abort.c:79
#2  0x00007fe99ea7c983 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007fe99ea828c6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007fe99ea819d9 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007fe99ea822d5 in __gxx_personality_v0 () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007fe99e863e33 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#7  0x00007fe99e8646b5 in _Unwind_Resume () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#8  0x00007fe99ee82887 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=<optimized out>, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#9  std::__shared_ptr<epics::pvAccess::Destroyable, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=<optimized out>, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
--Type <RET> for more, q to quit, c to continue without paging--
#10 std::shared_ptr<epics::pvAccess::Destroyable>::~shared_ptr (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr.h:103
#11 epics::pvAccess::Destroyable::cleaner::operator() (this=<optimized out>) at ../../src/utils/pv/destroyable.h:52
#12 0x000056302ca61dd7 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d5b12a0)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#13 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d5b12a0)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#14 0x00007fe99eeecdbb in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x56302d5b20a0, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#15 std::__shared_ptr<epics::pvAccess::ClientChannelImpl, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x56302d5b2098, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#16 std::shared_ptr<epics::pvAccess::ClientChannelImpl>::~shared_ptr (this=0x56302d5b2098, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr.h:103
#17 (anonymous namespace)::BaseRequestImpl::~BaseRequestImpl (this=0x56302d5b2040, __vtt_parm=<optimized out>, 
    __in_chrg=<optimized out>) at ../../src/remoteClient/clientContextImpl.cpp:217
#18 0x00007fe99eeef98d in (anonymous namespace)::ChannelGetImpl::~ChannelGetImpl (this=<optimized out>, 
    __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at ../../src/client/pv/pvAccess.h:494
#19 0x00007fe99eeef9f2 in (anonymous namespace)::ChannelGetImpl::~ChannelGetImpl (this=0x56302d5b2040, 
    __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at ../../src/remoteClient/clientContextImpl.cpp:608
#20 std::_Sp_counted_ptr<(anonymous namespace)::ChannelGetImpl*, (__gnu_cxx::_Lock_policy)2>::_M_dispose (
    this=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:377
#21 0x00007fe99eed33db in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d5b2320)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#22 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d5b2320)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#23 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fe99cb54d18, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr_base.h:728
#24 std::__shared_ptr<epics::pvAccess::TransportSender, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fe99cb54d10, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#25 std::shared_ptr<epics::pvAccess::TransportSender>::~shared_ptr (this=0x7fe99cb54d10, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr.h:103
#26 epics::pvAccess::detail::AbstractCodec::processSendQueue (this=0x7fe990000bf0) at ../../src/remote/codec.cpp:840
#27 0x00007fe99eed6808 in epics::pvAccess::detail::BlockingTCPTransportCodec::sendThread (this=0x7fe990000bf0)
--Type <RET> for more, q to quit, c to continue without paging--
    at ../../src/remote/codec.cpp:1169
#28 0x00007fe99ebfc43e in epicsThreadCallEntryPoint (pPvt=0x7fe990000d90) at ..//osi/epicsThread.cpp:103
#29 0x00007fe99ec0287b in start_routine (arg=0x7fe990005a90) at ..//osi/os/posix/osdThread.c:411
#30 0x00007fe99e678fa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#31 0x00007fe99e78b4cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95


Thread 2 (Thread 0x7fe99e3e0740 (LWP 27886)):
#0  0x00007fe99e67a495 in __GI___pthread_timedjoin_ex (threadid=140641348474624, 
    thread_return=thread_return@entry=0x7ffdb3485c90, abstime=abstime@entry=0x0, block=block@entry=true)
    at pthread_join_common.c:89
#1  0x00007fe99e67a26c in __pthread_join (threadid=<optimized out>, thread_return=thread_return@entry=0x7ffdb3485c90)
    at pthread_join.c:24
#2  0x00007fe99ec03e8c in epicsThreadMustJoin (id=0x56302d54fa00) at ..//osi/os/posix/osdThread.c:638
#3  0x00007fe99ebfc905 in epicsThread::exitWait (this=0x56302d54f850, delay=delay@entry=5) at ..//osi/epicsThread.cpp:191
#4  0x00007fe99eec3975 in epics::pvAccess::BlockingUDPTransport::close (this=0x56302d50f140, 
    waitForThreadToComplete=<optimized out>) at ../../src/remote/blockingUDPTransport.cpp:177
#5  0x00007fe99eef750c in (anonymous namespace)::InternalClientContextImpl::destroy (this=0x56302d4ecac0)
    at /usr/include/c++/8/bits/shared_ptr_base.h:1018
#6  0x00007fe99eee5e2c in epics::pvAccess::Destroyable::cleaner::operator() (this=<optimized out>)
    at ../../src/utils/pv/destroyable.h:54
#7  0x00007fe99eea92ed in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d4ed490)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#8  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d4ed490)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#9  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x56302d4ec478, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr_base.h:728
#10 std::__shared_ptr<epics::pvAccess::ChannelProvider, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x56302d4ec470, 
--Type <RET> for more, q to quit, c to continue without paging--
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#11 std::shared_ptr<epics::pvAccess::ChannelProvider>::~shared_ptr (this=0x56302d4ec470, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr.h:103
#12 pvac::ClientProvider::Impl::~Impl (this=0x56302d4ec470, __in_chrg=<optimized out>) at ../../src/client/client.cpp:239
#13 std::_Sp_counted_ptr<pvac::ClientProvider::Impl*, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr_base.h:377
#14 0x00007fe99eea53bf in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d4ec7b0)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#15 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x56302d4ec7b0)
    at /usr/include/c++/8/bits/shared_ptr_base.h:148
#16 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr_base.h:728
#17 std::__shared_ptr<pvac::ClientProvider::Impl, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=<optimized out>, 
    __in_chrg=<optimized out>) at /usr/include/c++/8/bits/shared_ptr_base.h:1167
#18 std::shared_ptr<pvac::ClientProvider::Impl>::~shared_ptr (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/include/c++/8/bits/shared_ptr.h:103
#19 pvac::ClientProvider::~ClientProvider (this=<optimized out>, __in_chrg=<optimized out>)
    at ../../src/client/client.cpp:285
#20 0x000056302ca5ed1d in main (argc=<optimized out>, argv=0x7ffdb3486228) at ../pvget.cpp:130
```

**Comment by mdavidsaver** [comment]
The server in question is experimental (not QSRV).

**Comment by mdavidsaver** [comment]
Looks like it is all servers (not surprising).  Candidate fix 37b7a0708fb34ba5d12833bfc71d78dcefda4d57 fixes the crashes I was seeing w/ pvget or pvput in a shell loop.
